featurebyte/service/feature_table_cache.py

```diff
@@ -1,111 +1,367 @@
+        serving_names_mapping: Optional[Dict[str, str]] = None,
     ) -> None:
         """
-        Update Feature Table Cache by adding new feature definitions.
+        Create or update feature table cache
 
         Parameters
         ----------
-        observation_table_id: PydanticObjectId
-            Observation table id
-        feature_definitions: List[CachedFeatureDefinition]
-            Feature definitions
+        feature_store: FeatureStoreModel
+            Feature Store object
+        observation_table: ObservationTableModel
+            Observation table object
+        graph: QueryGraph
+            Graph definition
+        nodes: List[Node]
+            Input node names
+        feature_list_id: Optional[PydanticObjectId]
+            Optional feature list id
+        serving_names_mapping: Optional[Dict[str, str]]
+            Optional serving names mapping if the observations set has different serving name columns
+            than those defined in Entities
         """
-        document = await self.get_document_for_observation_table(observation_table_id)
-        if document:
-            existing_features = {
-                feat.definition_hash: feat for feat in document.feature_definitions
-            }
-            for feature in feature_definitions:
-                if feature.definition_hash not in existing_features:
-                    existing_features[feature.definition_hash] = feature
-                else:
-                    existing = existing_features[feature.definition_hash]
-                    if existing.feature_id is None:
-                        existing_features[feature.definition_hash] = feature
-
-            await self.update_document(
-                document_id=document.id,
-                data=FeatureTableCacheUpdate(feature_definitions=list(existing_features.values())),
-                return_document=False,
+        is_feature_list_deployed = False
+        if feature_list_id:
+            try:
+                feature_list = await self.feature_list_service.get_document(feature_list_id)
+                is_feature_list_deployed = feature_list.deployed
+            except DocumentNotFoundError:
+                is_feature_list_deployed = False
+
+        cache_metadata = (
+            await self.feature_table_cache_metadata_service.get_or_create_feature_table_cache(
+                observation_table_id=observation_table.id,
             )
+        )
+        feature_table_cache_exists = bool(cache_metadata.feature_definitions)
+        cached_nodes, non_cached_nodes = await self.split_cached_and_non_cached_nodes(
+            cache_metadata, graph, nodes
+        )
+
+        db_session = await self.session_manager_service.get_feature_store_session(
+            feature_store=feature_store
+        )
+        if feature_table_cache_exists:
+            # if feature table cache exists - update existing table with new features
+            if non_cached_nodes:
```

Comment:
We can return early without creating the session if feature table cache exists and there is non-cached nodes.